File: services/langchain-agent/app/domain/repositories/agent_repository.py
"""
Agent repository interface.

This module defines the interface for agent repositories.
"""

from abc import ABC, abstractmethod
from typing import List, Optional, Dict, Any

from app.core.entities.agent import Agent, AgentType


class AgentRepository(ABC):
    """
    Agent repository interface.
    
    This interface defines the contract for repositories that manage agents.
    """
    
    @abstractmethod
    async def create(self, agent: Agent) -> Agent:
        """
        Create a new agent.
        
        Args:
            agent: The agent to create
            
        Returns:
            The created agent with ID and timestamps
        """
        pass
    
    @abstractmethod
    async def get_by_id(self, agent_id: str, user_id: Optional[str] = None) -> Optional[Agent]:
        """
        Get an agent by ID.
        
        Args:
            agent_id: The ID of the agent to retrieve
            user_id: Optional user ID for access control
            
        Returns:
            The agent if found, None otherwise
        """
        pass
    
    @abstractmethod
    async def update(
        self, 
        agent_id: str, 
        user_id: Optional[str] = None, 
        **kwargs
    ) -> Optional[Agent]:
        """
        Update an agent.
        
        Args:
            agent_id: The ID of the agent to update
            user_id: Optional user ID for access control
            **kwargs: Fields to update
            
        Returns:
            The updated agent if found, None otherwise
        """
        pass
    
    @abstractmethod
    async def delete(self, agent_id: str, user_id: Optional[str] = None) -> bool:
        """
        Delete an agent.
        
        Args:
            agent_id: The ID of the agent to delete
            user_id: Optional user ID for access control
            
        Returns:
            True if deleted, False otherwise
        """
        pass
    
    @abstractmethod
    async def list(
        self, 
        user_id: Optional[str] = None, 
        agent_type: Optional[AgentType] = None, 
        skip: int = 0, 
        limit: int = 10
    ) -> List[Agent]:
        """
        List agents with filtering and pagination.
        
        Args:
            user_id: Optional user ID for filtering
            agent_type: Optional agent type for filtering
            skip: Number of agents to skip
            limit: Maximum number of agents to return
            
        Returns:
            List of agents
        """
        pass
    
    @abstractmethod
    async def count(
        self, 
        user_id: Optional[str] = None, 
        agent_type: Optional[AgentType] = None
    ) -> int:
        """
        Count agents with filtering.
        
        Args:
            user_id: Optional user ID for filtering
            agent_type: Optional agent type for filtering
            
        Returns:
            Number of agents matching the filters
        """
        pass

File: services/frontend/src/domain/repositories/AgentRepository.js
/**
 * Repository interface for Agent entities.
 * This is an abstract interface that defines the contract for agent data access.
 * Implementations will be provided in the infrastructure layer.
 */
export class AgentRepository {
  /**
   * Get all agents for a user
   * @param {string} userId - User ID
   * @returns {Promise<Array>} Promise resolving to array of agents
   */
  async getAll(userId) {
    throw new Error('Method not implemented');
  }

  /**
   * Get agent by ID
   * @param {string} agentId - Agent ID
   * @param {string} userId - User ID for access control
   * @returns {Promise<Object>} Promise resolving to an agent
   */
  async getById(agentId, userId) {
    throw new Error('Method not implemented');
  }

  /**
   * Create a new agent
   * @param {Object} agentData - Agent data
   * @param {string} userId - User ID
   * @returns {Promise<Object>} Promise resolving to created agent
   */
  async create(agentData, userId) {
    throw new Error('Method not implemented');
  }

  /**
   * Update an agent
   * @param {string} agentId - Agent ID
   * @param {Object} agentData - Updated agent data
   * @param {string} userId - User ID for access control
   * @returns {Promise<Object>} Promise resolving to updated agent
   */
  async update(agentId, agentData, userId) {
    throw new Error('Method not implemented');
  }

  /**
   * Delete an agent
   * @param {string} agentId - Agent ID
   * @param {string} userId - User ID for access control
   * @returns {Promise<boolean>} Promise resolving to deletion success
   */
  async delete(agentId, userId) {
    throw new Error('Method not implemented');
  }

  /**
   * Execute an agent
   * @param {string} agentId - Agent ID
   * @param {Object} executionData - Execution parameters
   * @param {string} userId - User ID for access control
   * @returns {Promise<Object>} Promise resolving to execution details
   */
  async execute(agentId, executionData, userId) {
    throw new Error('Method not implemented');
  }
} 